### PR TITLE
Inventory favorites

### PR DESCRIPTION
**Pull request by akleshchev**

### CODE REVIEW

**Review by marchcat — APPROVED** [review]
Wow, this one is big.

I'd suggest building this change as a project viewer first in order to test it separately, And then, probably, just skip Maint C and merge it right into `develop`.

File: indra/newview/llappearancemgr.cpp [hunk 1]
```diff
@@ -2042,7 +2042,7 @@ bool LLAppearanceMgr::getCanReplaceCOF(const LLUUID& outfit_cat_id)
 }
 
 // Moved from LLWearableList::ContextMenu for wider utility.
-bool LLAppearanceMgr::canAddWearables(const uuid_vec_t& item_ids) const
+bool LLAppearanceMgr::canAddWearables(const uuid_vec_t& item_ids, bool warn_on_type_mismarch) const
```
Comment (maxim-productengine) [inline indra/newview/llappearancemgr.cpp:2045]: "mismarch" typo

File: indra/newview/llfloaterinventorysettings.cpp [hunk 2]
```diff
@@ -28,17 +28,38 @@
...
 {
-    getChild<LLButton>("ok_btn")->setCommitCallback(boost::bind(&LLFloater::closeFloater, this, false));
+    getChild<LLUICtrl>("favorites_color")->setCommitCallback(boost::bind(&LLFloaterInventorySettings::updateColorSwatch, this));
     return TRUE;
 }
 
+void LLFloaterInventorySettings::updateColorSwatch()
+{
+    bool val = getChild<LLUICtrl>("favorites_color")->getEnabled();
+    getChild<LLUICtrl>("favorites_color")->setEnabled(val);
```
Comment (maxim-productengine) [inline indra/newview/llfloaterinventorysettings.cpp:52]: Not sure what happens here.

File: indra/newview/llinventoryfunctions.cpp [hunk 3]
```diff
@@ -2420,6 +2420,123 @@ void ungroup_folder_items(const LLUUID& folder_id)
...
+    return object->getIsFavorite();
+}
+
+void set_favorite(const LLUUID& obj_id, bool favorite)
+{
+    LLInventoryObject* obj = gInventory.getObject(obj_id);
+
+    if (obj && obj->getIsLinkType())
+    {
+        obj = gInventory.getObject(obj_id);
```
Comment (maxim-productengine) [inline indra/newview/llinventoryfunctions.cpp:2517]: Should we use 'object->getLinkedUUID()' here as id?

File: indra/newview/llinventoryfunctions.cpp [hunk 4]
```diff
@@ -2420,6 +2420,123 @@ void ungroup_folder_items(const LLUUID& folder_id)
...
+        favorite_send(obj, obj_id, favorite);
+    }
+}
+
+void toggle_favorite(const LLUUID& obj_id)
+{
+    LLInventoryObject* obj = gInventory.getObject(obj_id);
+    if (obj && obj->getIsLinkType())
+    {
+        obj = gInventory.getObject(obj_id);
```
Comment (maxim-productengine) [inline indra/newview/llinventoryfunctions.cpp:2531]: Same question as above ('obj->getLinkedUUID()')

File: indra/newview/llinventoryfunctions.h [hunk 5]
```diff
@@ -325,6 +329,21 @@ class LLIsTypeWithPermissions : public LLInventoryCollectFunctor
     LLUUID          mGroupID;
 };
 
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// Class LLFavoritesCollector
+//
+// Simple class that collects calling cards that are not null, and not
```
Comment (maxim-productengine) [inline indra/newview/llinventoryfunctions.h:335]: No more calling cards collectors.

File: indra/newview/skins/default/xui/en/menu_gallery_inventory.xml [hunk 6]
```diff
@@ -125,6 +125,22 @@
...
+         parameter="favorite" />
+    </menu_item_call>
+  <menu_item_call
+   label="Remove from favorites"
+   layout="topleft"
+   name="favorites_remove">
+    <menu_item_call.on_click
+     function="Inventory.DoToSelected"
+     parameter="favorite" />
+  </menu_item_call>
```
Comment (maxim-productengine) [inline indra/newview/skins/default/xui/en/menu_gallery_inventory.xml:143]: Do we need these 2 menu items? We have similar 2 below.

**Review by maxim-productengine — CHANGES_REQUESTED** [review]
Mostly looks good, but probably couple minor changes are required.

**Comment by akleshchev** [comment]
Thank you! Will fix those.

File: indra/newview/llfloaterinventorysettings.cpp (hunk 2 above)
Comment (akleshchev) [reply indra/newview/llfloaterinventorysettings.cpp:52]: It should have been getChild<LLUICtrl>("favorites_swatch")->setEnabled(val);

File: indra/newview/skins/default/xui/en/menu_gallery_inventory.xml (hunk 6 above)
Comment (akleshchev) [reply indra/newview/skins/default/xui/en/menu_gallery_inventory.xml]: Removed, thank you.

File: indra/newview/llinventoryfunctions.cpp (hunk 3 above)
Comment (akleshchev) [reply indra/newview/llinventoryfunctions.cpp]: Fixed.

**Review by maxim-productengine — APPROVED** [review]
Ship it!

**Pull request merged**